src/helpers.cpp

```diff
@@ -151,4 +151,182 @@ Vec3 Triangulate2(const SE3 &g12, const Vec2 &xc1, const Vec2 &xc2) {
+  Vec3 t12{g12.T()};
+  Mat3 R12{g12.R()};
+  Mat3 R21{R12.transpose()};
+  Vec3 t21{-1 * R12.transpose() * t12};
+
+  // Create homogeneous coordinates
+  Vec3 f0{xc1(0), xc1(1), 1.0};
+  f0.normalize();
+  Vec3 f1{xc2(0), xc2(1), 1.0};
+  f1.normalize();  
+
+  Vec3 m0{R21 * f0};
+  Vec3 m1{f1};
+
+  float a0 = ((m0 / m0.norm()).cross(t21)).norm();
+  float a1 = ((m1 / m1.norm()).cross(t21)).norm();
+
+  Vec3 m0_prime;
+  Vec3 m1_prime;
+
+  if(a0 <= a1)
+  {
+    Vec3 n1 = m1.cross(t21);
+    Vec3 n1_hat = n1 / n1.norm();
+    m0_prime = m0 - (m0.dot(n1_hat)) * n1_hat;
+    m1_prime = m1;
+  }
+  else
+  {
+    Vec3 n0 = m0.cross(t21);
+    Vec3 n0_hat = n0 / n0.norm();
+    m0_prime = m0;
+    m1_prime = m1 - (m1.dot(n0_hat)) * n0_hat;
+  }
+
+  Vec3 z = m1_prime.cross(m0_prime);
+
+  Vec3 x = ((z.dot(t21.cross(m0_prime))) / pow(z.norm(),2)) * m1_prime;
+
+  return x;
+}
+
+
+Vec3 Triangulate4(const SE3 &g12, const Vec2 &xc1, const Vec2 &xc2) {
+
+  // Initalize the Rotation and Translation Matricies
+  Vec3 t12{g12.T()};
+  Mat3 R12{g12.R()};
+  Mat3 R21{R12.transpose()};
+  Vec3 t21{-1 * R12.transpose() * t12};
+
+  // Create homogeneous coordinates
+  Vec3 f0{xc1(0), xc1(1), 1.0};
+  f0.normalize();
+  Vec3 f1{xc2(0), xc2(1), 1.0};
+  f1.normalize();  
+
+  Vec3 m0{R21 * f0};
+  Vec3 m1{f1};
+
+  Vec3 m0_hat = m0 / m0.norm();
+  Vec3 m1_hat = m1 / m1.norm();
+
+  Eigen::Matrix<double, 3, 2> A;
+  A.row(0) << m0_hat(0), m1_hat(0);
+  A.row(1) << m0_hat(1), m1_hat(1);
+  A.row(2) << m0_hat(2), m1_hat(2);
+
+  Vec3 t21_hat = t21 / t21.norm();
+  Mat3 I = Eigen::Matrix3d::Identity();
+
+  Eigen::Matrix<double, 2, 3> B;
+  B = A.transpose() * (I - t21_hat * t21_hat.transpose());
+
+  Eigen::JacobiSVD<Eigen::Matrix<double, 2, 3>> svd(B, Eigen::ComputeFullV);
+  Mat3 V = svd.matrixV();
+  Vec3 n_prime_hat = V.col(1);
+
+  Vec3 m0_prime = m0 - m0.dot(n_prime_hat) * n_prime_hat;
+  Vec3 m1_prime = m1 - m1.dot(n_prime_hat) * n_prime_hat;
+
+  Vec3 z = m1_prime.cross(m0_prime);
+
+  Vec3 x = ((z.dot(t21.cross(m0_prime))) / pow(z.norm(),2)) * m1_prime;
+
+  return x;
+}
+
+Vec3 Triangulate5(const SE3 &g12, const Vec2 &xc1, const Vec2 &xc2) {
+
+  // Initalize the Rotation and Translation Matricies
+  Vec3 t12{g12.T()};
+  Mat3 R12{g12.R()};
+  Mat3 R21{R12.transpose()};
+  Vec3 t21{-1 * R12.transpose() * t12};
+
+  // Create homogeneous coordinates
+  Vec3 f0{xc1(0), xc1(1), 1.0};
+  f0.normalize();
+  Vec3 f1{xc2(0), xc2(1), 1.0};
+  f1.normalize();  
+
+  Vec3 m0{R21 * f0};
+  Vec3 m1{f1};
+
+  Vec3 m0_hat = m0 / m0.norm();
+  Vec3 m1_hat = m1 / m1.norm();
+
+  Vec3 n_a = (m0_hat + m1_hat).cross(t21);
+  Vec3 n_b = (m0_hat - m1_hat).cross(t21);
+
+  Vec3 n_prime_hat = n_a.norm() >= n_b.norm() ? n_a : n_b;
+
+  Vec3 m0_prime = m0 - m0.dot(n_prime_hat) * n_prime_hat;
+  Vec3 m1_prime = m1 - m1.dot(n_prime_hat) * n_prime_hat;
+
+  Vec3 z = m1_prime.cross(m0_prime);
+
+  // check_cheirality(z, t21, m1_prime, m0_prime);
```

Comment:
You have defined the interfaces to `check_cheirality`, `check_angular_reprojection`, and `check_parallax` to take intermediate variables of `Triangulate5` as input. However, these checks are quite general and could be used in more problems if their inputs were more general as well. For example, we can move these checks to a higher level (into `Feature::Triangulate`) and run the checks with every triangulation if the inputs were the outputs of triangulation.

Reply:
A couple of things - should I run these checks just for triangulation methods based on angular reprojection error ( ie ```Triangulate3```, ```Triangulate4```, and ```Triangulate5```) or should it be part of the ```Feature::Triangulate```?Moreover, if I make it general, it will have a few repeated operations, will it make this slower?

Reply:
My gut says the cheirality check and the parallax check should be run for all methods and the angular reprojection error check should only be run on the new methods that minimize angular reprojection error. My thinking is that finding a good threshold for maximum angular reprojection error might be hard for methods that aren't directly minimizing it. These are (small) design choices, though. I suspect that when the feature tracking is good enough, all methods will give small angular reprojection error.

If you make it general, it might be just a tiny bit slower. At the moment, we have compiler optimizations turned off -- see line 25 of `CMakeLists.txt`. But it wouldn't surprise me if compiler optimizations did some function inlining and got rid of the inefficiency.